spec/index.bs

```diff
@@ -6321,11 +6323,37 @@ dictionary GPURenderBundleEncoderDescriptor : GPUObjectDescriptorBase {
 # Queues # {#queues}
 
 <script type=idl>
+typedef [EnforceRange] unsigned long long GPUSignalValue;
+
+dictionary GPUTextureSubresourceRange {
+    GPUTextureAspect aspect = "all";
+    GPUIntegerCoordinate baseMipLevel = 0;
+    GPUIntegerCoordinate mipLevelCount;
+    GPUIntegerCoordinate baseArrayLayer = 0;
+    GPUIntegerCoordinate arrayLayerCount;
+};
+
+dictionary GPUTextureSubresources: GPUTextureSubresourceRange {
+    required GPUTexture texture;
+};
+
+dictionary GPUResourceTransfer {
+    required GPUQueue receivingQueue;
+    required sequence<GPUBuffer> buffers;
+    required sequence<GPUTextureSubresources> textureSubresources;
+};
+
 interface GPUQueue {
-    undefined submit(sequence<GPUCommandBuffer> commandBuffers);
+    GPUSignalValue signal();
+    readonly attribute GPUSignalValue lastSignaledValue;
+
+    Promise<undefined> onCompletion(optional GPUSignalValue value);
 
-    GPUFence createFence(optional GPUFenceDescriptor descriptor = {});
-    undefined signal(GPUFence fence, GPUFenceValue signalValue);
+    undefined wait(GPUQueue queue, optional GPUSignalValue value);
+
+    undefined submit(
+        sequence<GPUCommandBuffer> commandBuffers,
+        optional sequence<GPUResourceTransfer> transfers = []);
```

Comment:
I *think* what's meant here is: "Getting signal values and waiting on signals are separate queue operations, rather than rolled into the submit() call, because writeBuffer and writeTexture also submit work to the queue. This avoids having to extend all three functions (submit, writeBuffer, and writeTexture) to return signal values and be able to wait on signals."

> I would have expected that after `submit` returns control to the caller, the developer is free to use the resources in the transfer list on the receiving queue and the WebGPU implementation handles setting up the fences as needed to avoid undefined behavior. What is the advantage of having the developer do this themselves?

We shouldn't just implicitly insert a wait() on the receiving queue at whatever arbitrary time the app happens to issue the transferring submit on the sending queue. The app needs to have control over what point in the receiving queue the resources are received, otherwise the receiving queue can stall too early and go idle when it shouldn't have to.

Reply:
> We shouldn't just implicitly insert a wait() on the receiving queue at whatever arbitrary time the app happens to issue the transferring submit on the sending queue. The app needs to have control over what point in the receiving queue the resources are received, otherwise the receiving queue can stall too early and go idle when it shouldn't have to.

I agree it would be sub-optimal to insert a wait on the receiving queue at the time of the transferring submit.  Why can't we insert a wait the first time you submit a command list with the object to the receiving queue?  Wouldn't we need to run the same logic to validate the developer hasn't missed a wait? Missing a wait would lead to undefined behavior.

Reply:
> Why can't we insert a wait the first time you submit a command list with the object to the receiving queue?

Because that would make the graph of synchronization implicit. Missing a wait would cause a validation error and not UB. As said in the meeting I'll detail more what the thought process was for this proposal.